File: LeetCode-Contests/LeetCode-contest-238/5741.py
class Solution:
    def maxBuilding(self, n: int, restrictions: List[List[int]]) -> int:
        dp = [-1] * n
        for vec in restrictions:
            dp[vec[0] - 1] = vec[1]
        dp[0] = 0

        for i in range(n):
            if dp[i] != -1:
                dp[i] = min(dp[i-1] + 1, dp[i])
            else:
                dp[i] = dp[i-1] + 1

        res = 0
        for i in range(n-1, -1, -1):
            if i < n - 1 and dp[i+1] + 1 < dp[i]:
                dp[i] = dp[i+1] + 1
            if dp[i] > res:
                res = dp[i]
        return res

File: LeetCode-Contests/LeetCode-contest-238/5741.cc
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
  int maxBuilding(int n, vector<vector<int>> &restrictions) {
    vector<int> dp(n, -1);
    for (auto &p : restrictions)
      dp[p[0] - 1] = p[1];
    dp[0] = 0;

    for (int i = 1; i < n; ++i) {
      if (dp[i] != -1)
        dp[i] = min(dp[i - 1] + 1, dp[i]);
      else
        dp[i] = dp[i - 1] + 1;
    }
    int res = 0;
    for (int i = n - 1; i >= 1; --i) {
      if (i < n - 1 && dp[i + 1] + 1 < dp[i])
        dp[i] = dp[i + 1] + 1;
      if (dp[i] > res)
        res = dp[i];
    }
    return res;
  }
};
